test/unit/asrouter/CFRPageActions.test.js

```diff
@@ -790,6 +790,40 @@ describe("CFRPageActions", () => {
 
         delete fakeRecommendation.template;
       });
+      describe("simultaneous recommendations", () => {
+        it("should prevent a second message if one is currently displayed", async () => {
+          const secondMessage = {...fakeRecommendation, id: "second_message"};
+          let messageAdded = await CFRPageActions.addRecommendation(fakeBrowser, fakeHost, fakeRecommendation, dispatchStub);
+
+          assert.isTrue(messageAdded);
+          assert.deepInclude(CFRPageActions.RecommendationMap.get(fakeBrowser), {
+            id: fakeRecommendation.id,
+            host: fakeHost,
+            content: fakeRecommendation.content,
+          });
+
+          messageAdded = await CFRPageActions.addRecommendation(fakeBrowser, fakeHost, secondMessage, dispatchStub);
+          // Adding failed
+          assert.isFalse(messageAdded);
+          // First message is still there
+          assert.deepInclude(CFRPageActions.RecommendationMap.get(fakeBrowser), {
+            id: fakeRecommendation.id,
+            host: fakeHost,
+            content: fakeRecommendation.content,
+          });
+        });
+        it("should send impressions just for the first message", async () => {
```

Comment:
I think it would be more definitive to check the actual state of the last provider impression / to run the frequency cap function than just inspecting the messages that get sent, although that would probably be a lot more difficult.